sunpy/map/mapbase.py

```diff
@@ -458,41 +458,75 @@ def _validate(self):
-        width = self.shape[1]
-        height = self.shape[0]
-
-        if (x is not None) & (x > width-1):
-            raise ValueError("X pixel value larger than image width ({width!s}).".format(width=width))
-        if (x is not None) & (y > height-1):
-            raise ValueError("Y pixel value larger than image height ({height!s}).".format(height=height))
-        if (x is not None) & (x < 0):
-            raise ValueError("X pixel value cannot be less than 0.")
-        if (x is not None) & (y < 0):
-            raise ValueError("Y pixel value cannot be less than 0.")
-
-        scale = np.array([self.scale['x'], self.scale['y']])
-        crpix = np.array([self.reference_pixel['x'], self.reference_pixel['y']])
-        crval = np.array([self.reference_coordinate['x'], self.reference_coordinate['y']])
-        # FIXME: not used?!
-        coordinate_system = [self.coordinate_system['x'], self.coordinate_system['y']]
-        x,y = wcs.convert_pixel_to_data(self.shape, scale, crpix, crval, x = x, y = y)
+    @u.quantity_input(x=u.deg, y=u.deg)
+    def data_to_pixel(self, x, y, origin=0):
+        """
+        Convert a data (world) coordinate to a pixel coordinate by using
+        `~astropy.wcs.WCS.wcs_world2pix`.
+
+        Parameters
+        ----------
+
+        x : `~astropy.units.Quantity`
+            Data coordinate of the CTYPE1 axis. (Normally solar-x).
+
+        y : `~astropy.units.Quantity`
+            Data coordinate of the CTYPE2 axis. (Normally solar-y).
+
+        origin : int
+            Origin of the top-left corner. i.e. count from 0 or 1.
+            See `~astropy.wcs.WCS.wcs_world2pix` for more information.
+
+        Returns
+        -------
+
+        x : float
+            Pixel coordinate on the CTYPE1 axis.
+
+        y : float
+            Pixel coordinate on the CTYPE2 axis.
+        """
+        x, y = self.wcs.wcs_world2pix(u.Quantity([x, y])[None, :].to(u.deg).value, origin)[0]
 
         return x, y
 
+    def pixel_to_data(self, x, y, origin=0):
+        """
+        Convert a pixel coordinate to a data (world) coordinate by using
+        `~astropy.wcs.WCS.wcs_pix2world`.
+
+        Parameters
+        ----------
+
+        x : float
+            Pixel coordinate of the CTYPE1 axis. (Normally solar-x).
+
+        y : float
+            Pixel coordinate of the CTYPE2 axis. (Normally solar-y).
+
+        origin : int
+            Origin of the top-left corner. i.e. count from 0 or 1.
+            See `~astropy.wcs.WCS.wcs_pix2world` for more information.
+
+        Returns
+        -------
+
+        x : `~astropy.units.Quantity`
+            Coordinate of the CTYPE1 axis. (Normally solar-x).
+
+        y : `~astropy.units.Quantity`
+            Coordinate of the CTYPE2 axis. (Normally solar-y).
+        """
+
+        x, y = self.wcs.wcs_pix2world(np.array([[x, y]]), origin)[0]
+
+        # WCS always outputs degrees.
+        x *= u.deg
+        y *= u.deg
```

Comment:
For legacy reasons I don't know if we can change these values to be quantities now. But a new generalized WCS framework is in the works and we can definitely make sure that this returns quantities where appropriate.